aten/src/ATen/core/dynamic_type.cpp

```diff
@@ -78,15 +78,13 @@ DynamicType::~DynamicType() {
   arguments_.~Arguments();
 }
 
-std::shared_ptr<const DynamicType> DynamicType::create(const Type& other) {
+SingletonOrSharedTypePtr<const DynamicType> DynamicType::create(const Type& other) {
   if (auto dynRaw = other.castRaw<DynamicType>()) {
     TORCH_INTERNAL_ASSERT(
         !dynRaw->weak_from_this().expired(),
         "Error creating dynamic type instance not managed by shared_ptr: ",
         other.str());
-  }
-  if (auto dyn = other.cast<DynamicType>()) {
-    return dyn;
+    return SingletonTypePtr<const DynamicType>(dynRaw);
```

Comment:
Can one have some sort of static assert that guarantees that copy constructor is not invoked when casting `SingletonTypePtr` to `SingletonOrSharedTypePtr`

Reply:
making that conversion cheap is a core goal of SingletonOrSharedTypePtr. If somebody changes that they're breaking everything it's for. https://github.com/pytorch/pytorch/blob/4d2d833976a3616953aee3fb75892fc0c7988309/aten/src/ATen/core/jit_type_base.h#L187-L188

Reply:
Sure, so it would be nice to have static assert somewhere, to prevent someone accidentally breaking it